src/field_screen_effect.c

```diff
@@ -1388,3 +1400,237 @@ void FieldCB_RushInjuredPokemonToCenter(void)
+#define tOffsetY          data[5]
+#define tTimer            data[6]
+#define tDelay            data[15]
+
+static void UpdateStairsMovement(s16 speedX, s16 speedY, s16 *offsetX, s16 *offsetY, s16 *timer)
+{
+    struct Sprite *playerSprite = &gSprites[gPlayerAvatar.spriteId];
+    struct ObjectEvent *playerObjectEvent = &gObjectEvents[gPlayerAvatar.objectEventId];
+
+    if (speedY > 0 || *timer > 6)
+        *offsetY += speedY;
+
+    *offsetX += speedX;
+    (*timer)++;
+    playerSprite->x2 = *offsetX >> 5;
+    playerSprite->y2 = *offsetY >> 5;
+    if (playerObjectEvent->heldMovementFinished)
+        ObjectEventForceSetHeldMovement(playerObjectEvent, GetWalkInPlaceNormalMovementAction(GetPlayerFacingDirection()));
+}
+
+static void Task_StairWarp(u8 taskId)
+{
+    s16 * data = gTasks[taskId].data;
+    struct ObjectEvent *playerObjectEvent = &gObjectEvents[gPlayerAvatar.objectEventId];
+    struct Sprite *playerSprite = &gSprites[gPlayerAvatar.spriteId];
+
+    switch (tState)
+    {
+        case 0:
+            LockPlayerFieldControls();
+            FreezeObjectEvents();
+            CameraObjectFreeze();
+            tState++;
+            break;
+        case 1:
+            if (!ObjectEventIsMovementOverridden(playerObjectEvent) || ObjectEventClearHeldMovementIfFinished(playerObjectEvent))
+            {
+                if (tDelay != 0)
+                    tDelay--;
+                else
+                {
+                    TryFadeOutOldMapMusic();
+                    PlayRainStoppingSoundEffect();
+                    playerSprite->oam.priority = 1;
+                    ForceStairsMovement(tMetatileBehavior, &tSpeedX, &tSpeedY);
+                    PlaySE(SE_EXIT);
+                    tState++;
+                }
+            }
+            break;
+        case 2:
+            UpdateStairsMovement(tSpeedX, tSpeedY, &tOffsetX, &tOffsetY, &tTimer);
+            tDelay++;
+            if (tDelay >= 12)
+            {
+                WarpFadeOutScreen();
+                tState++;
+            }
+            break;
+        case 3:
+            UpdateStairsMovement(tSpeedX, tSpeedY, &tOffsetX, &tOffsetY, &tTimer);
+            if (!PaletteFadeActive() && BGMusicStopped())
+                tState++;
+            break;
+        default:
+            gFieldCallback = FieldCB_DefaultWarpExit;
+            WarpIntoMap();
+            SetMainCallback2(CB2_LoadMap);
+            DestroyTask(taskId);
+            break;
+    }
+}
+
+void DoStairWarp(u16 metatileBehavior, u16 delay)
+{
+    u8 taskId = CreateTask(Task_StairWarp, 10);
+    gTasks[taskId].tMetatileBehavior = metatileBehavior;
+    gTasks[taskId].tDelay = delay;
+    Task_StairWarp(taskId);
+}
+
+#undef tMetatileBehavior
+#undef tSpeedX
+#undef tSpeedY
+#undef tOffsetX
+#undef tOffsetY
+#undef tTimer
+#undef tDelay
+
+bool32 IsDirectionalStairWarpMetatileBehavior(u16 metatileBehavior, u8 playerDirection)
+{
+    switch (playerDirection)
```

Comment:
I personally wonder if this is worth a switch case, but that's subjective. If so: indentation

Reply:
[`874d4d0` (#5278)](https://github.com/rh-hideout/pokeemerald-expansion/pull/5278/commits/874d4d047002a7651a4767e145c8b71a98996a84)